### PR TITLE
pinctrl: mcux_rt1xxx: add pinctrl support for rt1xxx series

### PR DESCRIPTION
**Pull request by hakehuang**
pinctrl: nxp imxrt1xxx: add pinctrl support

demo on lpuart and i2s driver.

nxp rt1xxx series has different iomuxc so we add sepcial pincontrol bindings definitions.
and for gpr registers sets we rename former pinmux to gpr, and driver can access gpr directly demoed in i2s driver

tested on imxrt1060_evk board with hello_world

### CODE REVIEW

**Comment by hakehuang** [comment]
below warning in CI is a must for pinctrl
```
-:698: WARNING:NEW_TYPEDEFS: do not add new typedefs
#698: FILE: include/drivers/pinctrl/pinctrl_soc_mcux_rt_common.h:27:
+typedef struct pinctrl_soc_pin {

- total: 0 errors, 1 warnings, 658 lines checked
```

File: boards/arm/mimxrt1060_evk/mimxrt1060_evk-pinctrl.dtsi [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * Copyright (c) 2021, NXP
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <dt-bindings/pinctrl/mcux_rt-pinctrl.h>
+
+&pinctrl {
+	sai1_default: sai1_default {
+		IOMUXC_GPIO_AD_B1_09_SAI1_MCLK {
+			pinmux = <0x401F8120 0x3 0x401F858C 0x1 0x401F8310 0x1>;
```
Comment (danieldegrasse) [inline boards/arm/mimxrt1060_evk/mimxrt1060_evk-pinctrl.dtsi:11]: Maybe move the `inputOnfield` parameter in the `pinmux` property to be a configurable value instead of part of the pin definition? My reasoning here is that the the `inputOnfield` property is usually given as an argument to `IOMUXC_SetPinMux`, where as the rest of the constants defined here are not

File: dts/arm/nxp/nxp_rt-pinctrl.dtsi [hunk 2]
```diff
@@ -0,0 +1,13 @@
+/*
```
Comment (danieldegrasse) [inline dts/arm/nxp/nxp_rt-pinctrl.dtsi:1]: Why is this file necessary?

File: dts/arm/nxp/nxp_rt.dtsi [hunk 3]
```diff
@@ -231,9 +231,12 @@
 			status = "disabled";
 		};
 
-		iomuxc: iomuxc@401f8000 {
+		pinctrl: pinctrl@401f8000 {
+			compatible = "nxp,mcux-rt-pinctrl";
 			reg = <0x401f8000 0x4000>;
-			label = "PINMUX_0";
+			#address-cells = <1>;
```
Comment (danieldegrasse) [inline dts/arm/nxp/nxp_rt.dtsi:237]: Do we need `address-cells` and `size-cells`? The dts parser was giving me a warning that they were unused

File: include/drivers/pinctrl/pinctrl_soc_mcux_rt_common.h [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	uint32_t config_register;
+	uint32_t input_on_field;
+};
+
+typedef struct pinctrl_soc_pin {
+	struct pinctrl_soc_pinmux pinmux;
+	uint32_t pin_ctrl_flags;
+} pinctrl_soc_pin_t;
+
+#define MCUX_RT_HYS_SHIFT   16
```
Comment (danieldegrasse) [inline include/drivers/pinctrl/pinctrl_soc_mcux_rt_common.h:32]: These constants are defined in the device files in the HAL, see [here](https://raw.githubusercontent.com/zephyrproject-rtos/hal_nxp/master/mcux/mcux-sdk/devices/MIMXRT1062/MIMXRT1062.h)

**Review by danieldegrasse — COMMENTED** [review]
Thanks for getting this implemented @hakehuang! Just a note: I built a version of pinctrl for the RT1xxx based off this PR ([here's a link to the fork](https://github.com/NXPmicro/zephyr/tree/pinctrl-pin-nodes)), that uses pin nodes at the SOC level as opposed to pin groupings defined at the board level. After implementing that method of pinctrl, I think your implementation may be the cleaner method, but I'm interested to hear your thoughts.

File: boards/arm/mimxrt1060_evk/mimxrt1060_evk-pinctrl.dtsi (hunk 1 above)
Comment (hakehuang) [reply boards/arm/mimxrt1060_evk/mimxrt1060_evk-pinctrl.dtsi:11]: I assume that a general guide is that using dts to confingure what is required. if we make it as an argument, so we need put this value to driver?

File: dts/arm/nxp/nxp_rt.dtsi (hunk 3 above)
Comment (hakehuang) [reply dts/arm/nxp/nxp_rt.dtsi:237]: yes, they are not used for now, I just keep it here and maybe we need use this.

File: dts/arm/nxp/nxp_rt-pinctrl.dtsi (hunk 2 above)
Comment (hakehuang) [reply dts/arm/nxp/nxp_rt-pinctrl.dtsi:1]: we need have a pinctl defined. so I keep it here and in the boards dts, details will be added to this node

File: include/drivers/pinctrl/pinctrl_soc_mcux_rt_common.h (hunk 4 above)
Comment (hakehuang) [reply include/drivers/pinctrl/pinctrl_soc_mcux_rt_common.h:32]: right, I just see those bit pos are common for rt series, so I put it here, and using hal defines would be better.

**Comment by hakehuang** [comment]
> Thanks for getting this implemented @hakehuang! Just a note: I built a version of pinctrl for the RT1xxx based off this PR ([here's a link to the fork](https://github.com/NXPmicro/zephyr/tree/pinctrl-pin-nodes)), that uses pin nodes at the SOC level as opposed to pin groupings defined at the board level. After implementing that method of pinctrl, I think your implementation may be the cleaner method, but I'm interested to hear your thoughts.

@danieldegrasse , I leave some comments to your commit directly, and I will close this PR, and wait for yours

File: include/drivers/pinctrl/pinctrl_soc_mcux_rt_common.h (hunk 4 above)
Comment (danieldegrasse) [reply include/drivers/pinctrl/pinctrl_soc_mcux_rt_common.h:31]: does the DTS use these definitions directly? I thought it pulled from the yaml file for pinctrl, and that these constants would only be used in the C driver file

**Comment by danieldegrasse** [comment]
@hakehuang To clarify- I think your approach is probably best here, I just wanted to make my change known in case we wanted to go that direction. Do you feel that the other approach would work better for the RT parts? I prefer your approach because it feels like it will lend itself better to auto generation of DTS files

**Comment by hakehuang** [comment]
> @hakehuang To clarify- I think your approach is probably best here, I just wanted to make my change known in case we wanted to go that direction. Do you feel that the other approach would work better for the RT parts? I prefer your approach because it feels like it will lend itself better to auto generation of DTS files
@danieldegrasse 
from the https://docs.zephyrproject.org/latest/guides/pinctrl/index.html , there are two options to support dts, your implement option 2, and I am using option 1.
option 1:
board dts+ board_pinctrl dtsi + soc.h
option 2:
board dts + board pinctrl dtsi + soc dtsi

so the differences are:
1. whether to generate soc.h(currently I put defines directly into pinctrl dtsi as there is no soc.h in hal for dts_root in our hal) or generate soc dtsi
2.  goup pins maybe make the code easy to  maintain

@dleach02, @mmahadevan108  do you have any comments on this?

btw: I add you as code-signer, thanks for the ideas on inoutput, and header

File: include/drivers/pinctrl/pinctrl_soc_mcux_rt_common.h (hunk 4 above)
Comment (hakehuang) [reply include/drivers/pinctrl/pinctrl_soc_mcux_rt_common.h:31]: right, dts does not use this, update to use sdk macros.

**Comment by danieldegrasse** [comment]
> so the differences are:
>
> 1. whether to generate soc.h(currently I put defines directly into pinctrl dtsi as there is no soc.h in hal for dts_root in our hal) or generate soc dtsi
> 2. goup pins maybe make the code easy to maintain

I think that's correct, yeah. The primary differences are that the pins can no longer be grouped by peripheral, and that SOCs will have all pins defined in a dtsi file (as opposed to per board definitions). I agree that grouped pins are a big benefit to your method.

When I set out to implement my method, I felt it would better lend itself to auto generation of board level DTS files, but after comparing each approach I no longer believe that. An auto generation tool would likely be aware of the relevant registers used to configure each pin, but would not be aware of the specific DTS names we used for each pinmux node in my approach.

**Comment by hakehuang** [comment]
> When I set out to implement my method, I felt it would better lend itself to auto generation of board level DTS files, but after comparing each approach I no longer believe that. An auto generation tool would likely be aware of the relevant registers used to configure each pin, but would not be aware of the specific DTS names we used for each pinmux node in my approach.

@danieldegrasse so you mean like the IOMUXC_GPIO_AD_B1_09_SAI1_MCLK is not consist over each soc? so the board dts will suffer to re-implement them, and if in this case use a soc.h file would be more concise, as we only need update the board pinctrl.dtsi. am I right?

**Comment by danieldegrasse** [comment]
@hakehuang Exactly, when moving between SOCs (or implementing a custom board) having all pinmux possibilities defined at the soc level in a dtsi file would simplify hand-coding new board pinctrl dts files.

However, if auto generation of the board level dts files became an option, then your method would work best. The reason I think that is because an autogeneration tool would always be aware of the SOC level registers used for configuring pinctrl, but would not necessarily be aware  of the pinctrl node names in Zephyr. If the auto generation tool was made aware of the pinctrl node names defined at the soc level dtsi than my method would also work.

**Comment by hakehuang** [comment]
> If the auto generation tool was made aware of the pinctrl node names defined at the soc level dtsi than my method would also work.

@danieldegrasse , you mean IOMUXC_GPIO_AD_B1_09_SAI1_MCLK? this name is generated by the config tools for device header file, I guess those can be aligned in auto tools.

**Comment by hakehuang** [comment]
> @hakehuang yeah, values like IOMUXC_GPIO_AD_B1_09_SAI1_MCLK. I guess the core question is this: Are those names defined by SOC data the config tools use, or by the config tools themselves? If the names are defined by SOC data, then I think my approach will work without too much risk, since those names are unlikely to change. Otherwise, I think we should go with your method.

@danieldegrasse  those names are defined by config tools usually default value will not change. so I think you method will always work.

**Comment by danieldegrasse** [comment]
@hakehuang I've been working on this for a few days, and after some discussions with the config tools team I've opened #41810. The main differences are that a header file is used to define the IOMUXC constants for pinmux configuration, and the pinctrl node property names are changed to make future auto generation efforts of the board level DTSI file more viable.
